### ISSUE TITLE
Plugin assumes completely wrong AOL

### ISSUE THREAD

**Issue by Selbi182** [maven-nar/nar-maven-plugin#242]
Hey there,

so I've recently tried to include this plugin in my project, but so far had a big headache over how exactly it's supposed to be configured. Specifically, I'm struggling to make the damn thing know the correct AOL.

`$ mvn install
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building Maven Examples :: nar-maven-plugin JNI Project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- nar-maven-plugin:3.2.3:nar-validate (default-nar-validate) @ nar-maven-plugin-jni-example ---
[INFO] Using AOL: amd64-Windows-msvc
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.211 s
[INFO] Finished at: 2016-08-12T14:32:54+02:00
[INFO] Final Memory: 10M/309M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.2.3:nar-validate (default-nar-validate) on project nar-maven-plugin-jni-example: Cannot deduce version number from: -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException`

It assumes I'm running amd64-Windows-msvc, apparently, which isn't remotely correct. It should be x86.Windows.g++, as per [this page](http://maven-nar.github.io/aol.html). Unfortunately, I don't know where to find that configuration file, and the page makes no mention of it either, so I don't know how to manually override it.

Chances are this is an issue on my end, but I just don't know what's causing this.

**Comment by Selbi182** [comment]
Things got a little more interesting with another sample project (without JNI, just a plain C-File):

> $ mvn install
> [INFO] Scanning for projects...
> [WARNING]
> [WARNING] Some problems were encountered while building the effective model for com.github.maven-nar.its.nar:it0001-executable:nar:1.0-SNAPSHOT
> [WARNING] 'build.plugins.plugin.version' for com.github.maven-nar:nar-maven-plugin is missing. @ line 49, column 15
> [WARNING]
> [WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
> [WARNING]
> [WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
> [WARNING]
> [INFO]
> [INFO] ------------------------------------------------------------------------
> [INFO] Building NAR Executable Test 1.0-SNAPSHOT
> [INFO] ------------------------------------------------------------------------
> [INFO]
> [INFO] --- nar-maven-plugin:3.5.0:nar-validate (default-nar-validate) @ it0001-executable ---
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 1.327 s
> [INFO] Finished at: 2016-08-12T15:12:53+02:00
> [INFO] Final Memory: 12M/309M
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.5.0:nar-validate (default-nar-validate) on project it0001-executable: msvc.version not specified and no VS<Version>COMNTOOLS environment variable can be found -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

It still assumes I want to use MSVC, which is wrong, but at least the error looks less generic.

**Comment by ctrueden** [comment]
Have you tested with the latest `nar-maven-plugin` snapshot from master? There was a recent bug-fix commit which may address these problems for you: https://github.com/maven-nar/nar-maven-plugin/commit/ad48e49187d09ce291e8f2c3a8dcd5a1fd2b1ddb

**Comment by dproteus13** [comment]
Also, as a last resort, you can specify the AOL via the <aol> tag inside
the plugin's <configuration>, I think.  Not ideal by any means, but it
might be worth knowing as a backup...

**Comment by Selbi182** [comment]
Nope, none of these approaches worked. Cloned the recent SNAPSHOT to my machine (and verified its existence in the .m2 folder), but I got the following error:

> [ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate (default-nar-validate) on project nar-maven-plugin-jni-example:
> Execution default-nar-validate of goal com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate failed. NullPointerException -> [Help 1]

Interestingly to note here, it didn't suggest the wrong AOL this time. Instead, it suggested no AOL at all.

Using 3.2.0 gives me the exact same problem as the version I used to start this topic, unfortuantely.

As for the <aol> tag... Call me ignorant, but I don't seem to find any documentation how to use that tag properly. Tried to write my AOL using dashes and dots (i.e. "<aol>x86-Windows-g++</aol>" vs. "<aol>x86.Windows.g++</aol>"), but neither approaches worked. Once again, though, I didn't get the wrong AOL suggested, but instead none at all.

Error message:
>[ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate (default-nar-validate) on project nar-maven-plugin-jni-example:
>Execution default-nar-validate of goal com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate failed. NullPointerException -> [Help 1]

**Comment by dproteus13** [comment]
The null pointer exception is probably because I told you the wrong place
to put the `<aol>` tag, I'm sorry about that!

Per this discussion, it should be inside `<configuration> <linker>`  :
https://github.com/maven-nar/nar-maven-plugin/issues/56

Pardon if my markup is wrong, I'm mostly on BitBucket :(

**Comment by Selbi182** [comment]
Okay, at least I got a more sensible error message:

>[ERROR] Failed to execute goal com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate (default-nar-validate) on project nar-maven-plugin-jni-example:
> Unable to parse configuration of mojo com.github.maven-nar:nar-maven-plugin:3.5.1-SNAPSHOT:nar-validate for parameter aol: Cannot find 'aol' in class com.github.maven_nar.Linker -> [Help 1]

Was this correct?

```
<plugins>
      <plugin>
        <groupId>com.github.maven-nar</groupId>
        <artifactId>nar-maven-plugin</artifactId>
	<version>3.5.1-SNAPSHOT</version>
        <extensions>true</extensions>
        <configuration>
          <linker>
	    <aol>x86-Windows-g++</aol>
          </linker>
          <libraries>
            <library>
              <type>executable</type>
              <run>true</run>
            </library>
          </libraries>
        </configuration>
      </plugin>
    </plugins>
```

**Comment by dproteus13** [comment]
Good! Sorry I think I hadn't had it quite right again :-/

As to why you have to set the linker, I suspect it's  related to the fact that this is really not the way things are supposed to work. The AOL should be derived during the validate state,  and that process is probably what sets the linker.   I'll be honest though, I only touch Windows to get into a Linux VM, so I'm not much help in figuring out why that part isn't working :-/

**Comment by Selbi182** [comment]
Meanwhile, I use Linux once in a blue moon, so there's that. :laughing: 

And no need to apologize, I'm just glad I got help at all!

@ctrueden If you'd still like to have some information in regards to the auto-detection not working properly, I'll gladly assist you.

**Comment by ctrueden** [comment]
@Selbi182 wrote:
> It'd be cool if this was mentioned somewhere in the documentation. I had no idea to manually setup the AOL before this was mentioned.

Could you please add it to [the FAQ](https://github.com/maven-nar/nar-maven-plugin/wiki/Frequently_Asked_Questions)?
